Automattic/pocket-casts-ios
# Implement data retriever with cache

| 📘 Part of: #1848  | 
|:---:|

Implement transcripts local storage.

The cache is implemented using an URL cache with back of local storage path.

Some implementation considerations:

- We are using an ephemeral session so no other HTTP cache will be used
- We only fill the cache if there is a proper 200 response. 
- If we get a 200 we cache the response should we handle scenarios were a transcript is updated later on with new text? Should eject transcripts based on time or any HTTP headers?
- I'm only caching the data object returned from HTTP, in terms of use from my tests looks is good enough. We can try to cache the full parsed TranscriptModel object later on.

## To test

 Start the app
- Ensure that you have the `transcripts` and `newShowNotesEndpoint` FF enabled

### Test regular online scenario

- Start playing an episode with transcript supports. Ex:  One of this list: https://lists.pocketcasts.com/b852a088-ccee-4e4b-a3c5-4bb7fd700a02
- Open the full screen player
- Tap on the transcript shelf button 
- Check if the transcripts shows correctly
- Close the transcripts view
- Open it again
- Check that no loading indicator is view or if show goes away quickly
- Start playing another episode with transcripts
- Tap on transcript shelf button. You should see the loading indicator
- Close transcript
- Open the transcript again and see if it's quickly load, no loading view
- Start playing the previous episode that should be already cached
- Tap on transcript shelf
- Check if the the transcript is show quickly

### Test offline scenario

- Go offline
- For the previous podcasts you opened, open the episode again and see if you can see the transcript

### Test auto-downloads

- Enable auto-downloads for UpNext Queue
- Add some episodes with Transcript to UpNext Queue
- Wait for the downloads to happen
- Go offline
- Open one of the episode donwloaded
- Check that you can see the transcript

## Checklist
